soes/linux/kernel-lan9252/spi-lan9252.c

```diff
@@ -0,0 +1,422 @@
+#define FAST_READ_DUMMY	1
+
+struct mutex lock;
+static DEFINE_MUTEX(lan9252_mutex);
+
+dev_t id;
+static struct device *dev;
+static struct class *lan9252_class;
+static struct spi_device *lan9252_devices;
+
+static int lan9252_ethercat_probe(struct spi_device *spi);
+static int lan9252_ethercat_remove(struct spi_device *spi);
+
+
+static int lan9252_open(struct inode *inode, struct file *filp)
+{
+	dev_dbg(dev, "%s() called\n", __FUNCTION__);
+	return 0;
+}
+
+static int lan9252_release(struct inode *inode, struct file *filp)
+{
+	dev_dbg(dev, "%s() called\n", __FUNCTION__);
+	return 0;
+}
+
+static loff_t lan9252_llseek(struct file *file, loff_t address,
+                         int whence)
+{
+	int status;
+
+	LAN9252_LOCK;
+
+        file->f_pos = address;
+
+	status = file->f_pos;
+
+	LAN9252_UNLOCK;
+	
+	return status;
+}
+
+static int * lan9252_reg_addr_make(int address)
+{
+
+	int temp;
+	static int lan9252_reg_addr[2];
+#ifdef DEBUG
+	int i;
+#endif
+
+	dev_dbg(dev, "%s() called\n", __FUNCTION__);
+
+	temp = address & 0xff00;
+	lan9252_reg_addr[0] = temp >> 8;
+	temp = address & 0x00ff;
+	lan9252_reg_addr[1] = temp;
+
+#ifdef DEBUG
+	for (i=0; i<2; i++)
+	modp
+		dev_dbg(dev, "%s() lan9252_reg_addr[%d] = 0x%x\n",
+		        __FUNCTION__, i, lan9252_reg_addr[i]);
+#endif
+   
+	return lan9252_reg_addr;
+}
+
+/*
+        int lan9252_test(void);
+        Description :   Check the connection status of LAN9252    
+        Retrun Value
+        0       :   success
+        other   :   fail   
+
+        Register Type           :   SYSTEM CONTROL AND STATUS REGISTERS
+        Register Name(SYMBOL)   :   BYTE ORDER TEST REGISTER(BYTE_TEST)
+        Offset                  :   0x0064
+        Size                    :   32bit
+        Default                 :   0x87654321
+        Datasheet description   :   Chip-level reset/configuration
+                                completion can be determined by first
+                                polling the 
+                                Byte Order Test Register(BYTE_TEST).
+*/
+static int lan9252_test(void)
+{
+        u8 command[4];
+        u8 test_buffer[4];
+        int status=0;
+#ifdef DEBUG
+	int i;
+#endif
+        command[0]=FAST_READ;
+        command[1]=0x00;
+        command[2]=0x64;
+        command[3]=FAST_READ_DUMMY;
+
+
+        status = spi_write_then_read(lan9252_devices, &command,
+                sizeof(command), test_buffer, sizeof(test_buffer));
+        if (status < 0 ) {
+                dev_err(dev, "%s() spi_write_then_read failed "
+                        "status=%d\n", __FUNCTION__, status);
+                return -1;
+        }
+
+        status = spi_write_then_read(lan9252_devices, &command,
+                sizeof(command), test_buffer, sizeof(test_buffer));
+        if (status < 0 ) {
+                dev_err(dev, "%s() spi_write_then_read failed "
+                        "status=%d\n", __FUNCTION__, status);
+                return -1;
+        }
```

Comment:
lan9252_test() checks whether lan9252 is connected from lan9252_probe().
For the first spi communication, strangely there was no response to the first read.
I do not know exactly why.

lan9252_test() When called the first time (lan9252_probe()), so we called twice.